### DISCUSSION TITLE
Using FluidPropertiesMaterialPT in a THM model

### DISCUSSION THREAD

**Discussion by amerallaf** [idaholab/moose#27692] [Q&A Modules: General]
Hello,

I am trying to make my fluid properties as a function of temperature-pressure, based on the example given in the following link:
https://mooseframework.inl.gov/source/fluidproperties/TemperaturePressureFunctionFluidProperties.html.

I am simply using the Heattransfer - THM coupled model that is presented in the following link:
https://mooseframework.inl.gov/source/actions/CoupledHeatTransferAction.html

I am simplying trying to make the fluid density value as a function of temperature-pressure.
However, I getting the following error:
```
thm0: *** ERROR ***
thm0: The following error occurred in the MaterialBase 'pipe1:fp_mat' of type ADFluidProperties3EqnMaterial.
thm0:
thm0: The declared AD material property 'e' of type 'double'
thm0: is already retrieved or declared as a non-AD property of type 'double'.
```
Which I am not really able to understand what it is.
The sub input is presented below (the master file is the same as in https://mooseframework.inl.gov/source/actions/CoupledHeatTransferAction.html).

I would appreciate your help and time in resolving this.

```c
# This is a part of T_wall_action test.

[GlobalParams]
  initial_p = 1.e5
  initial_vel = 0.
  initial_T = 300.

  closures = simple_closures
[]

[Functions]
  [rho]
    type = ParsedFunction
    symbol_names = 'TT pp'
    symbol_values = 'temperature pressure'
    expression = '1.5e3 + 0.13 * TT - 1.5e-4 * pp'
  []
[]

[AuxVariables]
  [Hw]
    family = monomial
    order = constant
    block = pipe1
  []
  [temperature]
    initial_condition = 300.
  []
  [pressure]
    initial_condition = 253727.1
  []
[]

[AuxKernels]
  [Hw_ak]
    type = ADMaterialRealAux
    variable = Hw
    property = 'Hw'
  []
  [./rho]
    type = MaterialRealAux
    variable = rho
    property = density
  [../]
[]

[FluidProperties]
  [eos]
    type = StiffenedGasFluidProperties
    gamma = 2.35
    q = -1167e3
    q_prime = 0
    p_inf = 1.e9
    cv = 1816
  []
[]

[Materials]
  [to_vars]
    type = FluidPropertiesMaterialPT
    pressure = pressure
    temperature = temperature
    fp = eos
  []
[]

[Closures]
  [simple_closures]
    type = Closures1PhaseSimple
  []
[]

[UserObjects]
  [T_uo]
    type = LayeredAverage
    direction = y
    variable = T
    num_layers = 10
    block = pipe1
  []
  [Hw_uo]
    type = LayeredAverage
    direction = y
    variable = Hw
    num_layers = 10
    block = pipe1
  []
[]

[Components]
  [pipe1]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '0 1 0'
    length = 1
    n_elems = 10
    A   = 1.28584e-01
    D_h = 8.18592e-01
    f = 0.01
    fp = eos
  []

  [hxconn]
    type = HeatTransferFromExternalAppTemperature1Phase
    flow_channel = pipe1
    Hw = 10000
    P_hf = 6.28319e-01
    initial_T_wall = 300.
    var_type = elemental
  []

  [inlet]
    type = InletMassFlowRateTemperature1Phase
    input = 'pipe1:in'
    m_dot = 10
    T = 400
  []

  [outlet]
    type = Outlet1Phase
    input = 'pipe1:out'
    p = 1e5
  []
[]

[Preconditioning]
  [pc]
    type = SMP
    full = true
  []
[]

[Postprocessors]
  [T_wall_avg]
    type = ElementAverageValue
    variable = T_wall
    execute_on = 'INITIAL TIMESTEP_END'
  []

  [htc_avg]
    type = ElementAverageValue
    variable = Hw
    execute_on = 'INITIAL TIMESTEP_END'
  []

  [T_avg]
    type = ElementAverageValue
    variable = T
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'
  dt = 0.1
  dtmin = 1e-7
  abort_on_solve_fail = true

  solve_type = 'NEWTON'
  line_search = 'basic'
  nl_rel_tol = 1e-7
  nl_abs_tol = 1e-4
  nl_max_its = 20

  l_tol = 1e-3
  l_max_its = 300

  start_time = 0.0

  petsc_options_iname = '-pc_type'
  petsc_options_value = ' lu'
[]

[Outputs]
  [out]
    type = Exodus
    show = 'T_wall T Hw'
  []
[]
```

**Comment by GiudGiud** [comment]
hello

the FluidPropertiesMaterialPT and the ADFluidProperties3EqnMaterial are conflicting because they declare properties with the same names

do you really need the first one? This is just for output usually?
If so, then add a suffix to the properties if possible

**Reply by amerallaf** [reply]
Hello @GiudGiud, thank you so much for your help.

Just to clarify, which properties that I am having with the same names that are causing the issue?
Also what do you mean by "the first one?", I would appreciate it if you can guide me in the script.

**Reply by GiudGiud** [reply]
the first one is `FluidPropertiesMaterialPT`. Do you really need this one? I only use it for output

Seems to be `e`, the internal energy, that is duplicated between both materials

**Reply by amerallaf** [reply]
Oh thank you for the clarification.

It seems to me that such practice (making density as a function of P,T) always requires to have:

```c
[Materials]
  [to_vars]
    type = FluidPropertiesMaterialPT
    pressure = pressure
    temperature = temperature
    fp = eos
  []
[]
``` 
To define the temperature and the pressure variables of the fluid, otherwise I don't care much about having it.
Is there another way you would recommend in this case?

**Reply by GiudGiud** [reply]
`StiffenedGasFluidProperties` should already have density as a function of P and T. are you trying to override that dependence?
I am asking this because I see a `rho` function in your input

**Reply by GiudGiud** [reply]
density is one of the variables btw, so you are not computing it directly from P and T, it is being solved for

Two ways forward:
- delete FluidPropertiesMaterialPT from your input and adapt the rest of the input to less properties being defined
- modify FluidPropertiesMaterialPT to be able to not compute 'e'. There's already such a modification in the code for entropy and the sound speed. it s a simple change

**Reply by amerallaf** [reply]
Yes I want to have my own function, and then ultimately do that for more variables (k, mu, etc.).

**Reply by GiudGiud** [reply]
Ideally you would use this
https://mooseframework.inl.gov/source/fluidproperties/TemperaturePressureFunctionFluidProperties.html
but it will take quite a bit of coding to get it to work with the thermal hydraulics module

Maybe instead of adding a 'rho' function, set parameters in `StiffenedGasFluidProperties` to make density close to what you need

@joshuahansel thoughts ?

**Reply by amerallaf** [reply]
Yes, I was working on that 
```c
    type = TemperaturePressureFunctionFluidProperties
```  
at first, because that would be better to estimate of NaK coolant properties.
Here I am trying to have rho, k, and mu as function of T,P, and I was getting the same error.

Error:

thm0: *** ERROR ***
thm0: The following error occurred in the MaterialBase 'pipe1:fp_mat' of type ADFluidProperties3EqnMaterial.
thm0:
thm0: The declared AD material property 'e' of type 'double'
thm0: is already retrieved or declared as a non-AD property of type 'double'.

Script:

```c
# This is a part of T_wall_action test.

cv = 4000

[GlobalParams]
  initial_p = 1.e5
  initial_vel = 0.
  initial_T = 300.

  closures = simple_closures
[]

[AuxVariables]
  [Hw]
    family = monomial
    order = constant
    block = pipe1
  []
  [temperature]
    initial_condition = 300.
  []
  [pressure]
    initial_condition = 253727.1
  []
[]

[AuxKernels]
  [Hw_ak]
    type = ADMaterialRealAux
    variable = Hw
    property = 'Hw'
  []
[]

[Functions]
  # This demonstrates how to define fluid properties that are functions
  # of the LOCAL value of the (p,T) variables
  # x for temperature
  # y for pressure
  [k]
    type = ParsedFunction
    expression = '14 + 1e-2 * x + 1e-5 * y'
  []
  [rho]
    type = ParsedFunction
    expression = '1.5e3 + 0.13 * x - 1.5e-4 * y'
  []
  [mu]
    type = ParsedFunction
    expression = '1e-3 + 2e-6 * x - 3e-9 * y'
  []
[]

[FluidProperties]
  [nak]
    type = TemperaturePressureFunctionFluidProperties
    cv = ${cv}
    k = k
    rho = rho
    mu = mu
  []
[]

[Materials]
  [to_vars]
    type = FluidPropertiesMaterialPT
    fp = nak
    outputs = 'all'
    output_properties = 'density k cp cv viscosity e h'
    pressure = pressure
    temperature = temperature

    compute_entropy = false
    compute_sound_speed = false
  []
[]

[Closures]
  [simple_closures]
    type = Closures1PhaseSimple
  []
[]

[UserObjects]
  [T_uo]
    type = LayeredAverage
    direction = y
    variable = T
    num_layers = 10
    block = pipe1
  []
  [Hw_uo]
    type = LayeredAverage
    direction = y
    variable = Hw
    num_layers = 10
    block = pipe1
  []
[]

[Components]
  [pipe1]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '0 1 0'
    length = 1
    n_elems = 10
    A   = 1.28584e-01
    D_h = 8.18592e-01
    f = 0.01
    fp = nak
  []

  [hxconn]
    type = HeatTransferFromExternalAppTemperature1Phase
    flow_channel = pipe1
    Hw = 10000
    P_hf = 6.28319e-01
    initial_T_wall = 300.
    var_type = elemental
  []

  [inlet]
    type = InletMassFlowRateTemperature1Phase
    input = 'pipe1:in'
    m_dot = 10
    T = 400
  []

  [outlet]
    type = Outlet1Phase
    input = 'pipe1:out'
    p = 1e5
  []
[]

[Preconditioning]
  [pc]
    type = SMP
    full = true
  []
[]

[Postprocessors]
  [T_wall_avg]
    type = ElementAverageValue
    variable = T_wall
    execute_on = 'INITIAL TIMESTEP_END'
  []

  [htc_avg]
    type = ElementAverageValue
    variable = Hw
    execute_on = 'INITIAL TIMESTEP_END'
  []

  [T_avg]
    type = ElementAverageValue
    variable = T
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'
  dt = 0.1
  dtmin = 1e-7
  abort_on_solve_fail = true

  solve_type = 'NEWTON'
  line_search = 'basic'
  nl_rel_tol = 1e-7
  nl_abs_tol = 1e-4
  nl_max_its = 20

  l_tol = 1e-3
  l_max_its = 300

  start_time = 0.0

  petsc_options_iname = '-pc_type'
  petsc_options_value = ' lu'
[]

[Outputs]
  [out]
    type = Exodus
    show = 'T_wall T Hw'
  []
[]
```

**Reply by GiudGiud** [reply]
the error is not coming from the FluidProperties, it s coming from adding this `FluidPropertiesMaterialPT`

**Reply by joshuahansel** [reply]
Right, so your first step is to make a FluidProperties object that you're happy with. Assuming that your relationships do not fit well with `StiffenedGasFluidProperties`, make your `TemperaturePressureFunctionFluidProperties` object and then try running with it in THM. As Guillaume hinted to, there might need to be additional development for this to work (because THM requires some minimum set of interfaces), but we should be able to add whatever interfaces are missing.

As Guillaume says, remove `FluidPropertiesMaterialPT` from your input file - THM does not want that.

**Comment by BrankoK89** [comment]
Hello, wanted to post here as the question seems to be related! I am trying to use piecewise function for flibe properties in THM, but it seems the type I am using is getting rejected:
```
[Functions]
  [./time_stepper]
        type = PiecewiseLinear
        x    = '-100000.0 -99999.0 -99990.0 -99980.0 -99900.0 -90000.0 0.0'
        y    = '0.5 1.0 1.0 10.0 50.0 100.0 100.0'
  [../]

  [./fuel_salt_rho_func] # density vs T
    type = PiecewiseLinear
    x = '750       2000'
    y = '2792.0210 1953.6460'
  [../]

  [./fuel_salt_enthalpy_func] # enthalpy vs T
    type = PiecewiseLinear
    x = '750        2000'
    y = '1.4130E+06 3.7680E+06'
  [../]

  [./fuel_salt_mu_func] # viscosity vs T
    type = PiecewiseLinear
    x = '750  760  770  780  790  800  810  820
            830  840  850  860  870  880  890  900
            910  920  930  940  950  960  970  980
            990  1000 1010 1020 1030 1040 1050 1060
            1070 1080 1090 1100 1110 1120 1130 1140
            1150 1160 1170 1180 1190 1200 2000'
    y = '2.7378E-02 2.5371E-02 2.3557E-02 2.1915E-02 2.0424E-02 1.9069E-02 1.7834E-02 1.6706E-02
            1.5674E-02 1.4728E-02 1.3859E-02 1.3060E-02 1.2324E-02 1.1645E-02 1.1017E-02 1.0436E-02
            9.8976E-03 9.3976E-03 8.9328E-03 8.5001E-03 8.0969E-03 7.7206E-03 7.3690E-03 7.0402E-03
            6.7322E-03 6.4434E-03 6.1724E-03 5.9178E-03 5.6783E-03 5.4529E-03 5.2404E-03 5.0400E-03
            4.8508E-03 4.6720E-03 4.5029E-03 4.3428E-03 4.1911E-03 4.0473E-03 3.9109E-03 3.7813E-03
            3.6582E-03 3.5411E-03 3.4297E-03 3.3235E-03 3.2224E-03 3.1259E-03 3.1259E-03'
  [../]
[]

[EOS]  #
  [./fuelsalt]
    type = PTFunctionsEOS #PTFunctionsEOS
    rho = fuel_salt_rho_func
    mu = fuel_salt_mu_func
    enthalpy = fuel_salt_enthalpy_func
    cp = 1884          # or your best estimate
    k = 1.0            # thermal conductivity in W/m/K
  [../]

  [./hx_salt_eos]
        type = SaltEquationOfState
        salt_type = Flibe
  [../]
[]
*** ERROR ***                                                                                                                                                                     
/home/branko89/cardinal/thm_toy.i:86.1: section '[EOS]' does not have an associated "Action".                                                                                     
 Common causes:                                                                                                                                                                   
- you misspelled the Action/section name                                                                                                                                          
- the app you are running does not support this Action/syntax                                                                                                                     
/home/branko89/cardinal/thm_toy.i:87.3: section '[EOS/fuelsalt]' does not have an associated "Action".                                                                            
 Common causes:                                                                                                                                                                   
- you misspelled the Action/section name                                                                                                                                          
- the app you are running does not support this Action/syntax                                                                                                                     
/home/branko89/cardinal/thm_toy.i:96.3: section '[EOS/hx_salt_eos]' does not have an associated "Action".                                                                         
 Common causes:                                                                                                                                                                   
- you misspelled the Action/section name                                                                                                                                          
- the app you are running does not support this Action/syntax   
- ```

**Reply by joshuahansel** [reply]
The fluid properties type is not being rejected - it's the name of your block: change `EOS` to `FluidProperties`. I think `EOS` may be SAM syntax.

**Reply by BrankoK89** [reply]
Yeah, I changed that, but then I get 
*** ERROR ***                                                                                                                                                                     
A 'PTFunctionsEOS' is not a registered object.                                                                                                                                    
                                                                                                                                                                                  
If you are trying to find this object in a dynamically loaded library, make sure that                                                                                             
the library can be found either in your "Problem/library_path" parameter or in the                                                                                                
MOOSE_LIBRARY_PATH environment variable.    
 So both are probably from SAM, what would MOOSE THM equivalent be?

**Reply by GiudGiud** [reply]
the object is an EOS too. It's from SAM.
that's why it s not registered in MOOSE

you have to use FluidProperties from the FLuidProperties module.

https://mooseframework.inl.gov/modules/fluid_properties/index.html

We have the TemperaturePressureFunctionFP that should be similar to what you are using

**Comment by BrankoK89** [comment]
I gave up on trying to adjust FLibeFluidProperties to work in my case in THM, so I wanted to switch to functions. Is that possible in MOOSE?

**Reply by GiudGiud** [reply]
you can try TemperaturePressureFunctionFP but that won't work in THM for the same reasons. It needs more routines implemented (speed of sound notably)

**Comment by BrankoK89** [comment]
In FlibeFluidProperties I implemented c_from_v_e and e_from_v_h, but I have issues reaching convergence. I tried adjusting drho_dp, as the object description suggests to improve that. If I leave drho_dp= 1.7324E-7 and sweep the system on specific volume values between 0.00049 and 0.000515 (1940 to 2040 kg/m3 density) the pressure values go from -6E8  to 6E8 which makes for some tough convergence. If I adjust drho_dp= 1E-5, that swing is more like from -4E6  to 4E6 making the convergence at least proceed to some point. I am unsure whether relaxing drho_dp would get me to where I need to be or if this EOS is just not suitable for THM (density being linear in p and T).

**Reply by GiudGiud** [reply]
> (density being linear in p and T

the density depending on pressure should help.
Maybe there are mistakes in those new FPs, maybe the initialization is too far from the solution for the nonlinear solver to converge

One way we could be more sure would be to use this FP and tune it for your fluid. It is known to work for THM once tuned. People have even tuned it to sodium and got convergence. 
https://mooseframework.inl.gov/source/fluidproperties/StiffenedGasFluidProperties.html

**Reply by BrankoK89** [reply]
So that would involve adjusting the input parameters to get to a point where density behaves close to literature for given P and T?

**Reply by GiudGiud** [reply]
yes exactly

**Reply by BrankoK89** [reply]
What is the best way to go about tuning them? I am attempting to transfer the StiffenedGasFluidProperties.C into Python functions. Then I can optimize the input parameters to deliver P, T, rho... close to what literature gives for FLiBe. I wouldn't know how to do such optimization in MOOSE.

**Reply by GiudGiud** [reply]
@joshuahansel or @licharlot have done this before iirc

**Reply by joshuahansel** [reply]
I wouldn't bother doing anything as formal as solving an optimization problem. Just use some experimental points and hand-tune the values to get a decent match to those points, and then see how that does in your simulation before spending more time on it.

**Reply by BrankoK89** [reply]
That was my intent, but working through a Python script would allow me to test the behavior over a range of interest for these properties and check whether they match with experimental values, rather than taking a stab at a few values. So I was wondering if there is a MOOSE way of doing this. Or should I build a very simple MOOSE/THM model and just play around with the parameters?

**Reply by licharlot** [reply]
You could use the FluidPropertiesInterrogator to get the propertie for given conditions.
Below is an example from the test suite:
```
 [FluidPropertiesInterrogator]
  fp = fp
  p = 1e5
  T = 300
[]

[FluidProperties]
  [./fp]
    type = IdealGasFluidProperties
    gamma = 1.4
    molar_mass = 0.02900055737704918
    mu = 1.823e-05
    k = 0.02568
  [../]
[]
```

**Reply by joshuahansel** [reply]
Yeah, I'd obviously start with the things we definitely know, like molar mass. Thermal conductivity and dynamic viscosity are constant in this EOS, so choose a median value for your range of interest. Then the other DoFs become relevant for various properties, so that's where you might choose values to fit at one experiment point, then plot out a thermodynamic space or curve to see if various properties like density are reasonable.

**Reply by joshuahansel** [reply]
Just saw Lise's suggestion - that's probably better. Test the properties at some boundaries of your application space (that'll be less effort than making plots).

**Reply by BrankoK89** [reply]
It seems the main parameter that I need to adjust is p_inf, the stiffness parameter. The value needs to be very high to attain flibe density.

**Comment by BrankoK89** [comment]
Yes, that FluidPropertiesInterrogator should make things much simpler! Thank you!
